enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/util/SOAPUtils.java

```diff
@@ -0,0 +1,71 @@
+ *
+ * WSO2 Inc. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.choreo.connect.enforcer.util;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.wso2.choreo.connect.enforcer.commons.logging.ErrorDetails;
+import org.wso2.choreo.connect.enforcer.commons.logging.LoggingConstants;
+import org.wso2.choreo.connect.enforcer.constants.APIConstants;
+
+import javax.xml.soap.MessageFactory;
+import javax.xml.soap.SOAPEnvelope;
+import javax.xml.soap.SOAPFault;
+import javax.xml.soap.SOAPMessage;
+import javax.xml.soap.SOAPPart;
+
+/**
+ * SOAPUtils class provides methods required for generating soap1.1 & soap1.2 messages.
+ * Currently, this only provides methods for generating fault messages.
+ */
+public class SOAPUtils {
+    private static final Logger log = LogManager.getLogger(SOAPUtils.class);
+
+    /**
+     * Returns a soap fault response message
+     *
+     * @param soapProtocolVersion 'SOAP 1.1 Protocol' or 'SOAP 1.2 Protocol'
+     * @param message             message text
+     * @param description         description of the fault
+     * @param code                response code
+     * @return xml formatted SOAP fault message as a String
+     */
+    public static String getSoapFaultMessage(String soapProtocolVersion,
+                                             String message, String description, String code) {
+        try {
+            MessageFactory factory = MessageFactory.newInstance(soapProtocolVersion);
+            SOAPMessage soapMsg = factory.createMessage();
+            SOAPPart part = soapMsg.getSOAPPart();
+            SOAPEnvelope envelope = part.getEnvelope();
+            SOAPFault soapFault = envelope.getBody().addFault();
+            if (soapProtocolVersion.equals(APIConstants.SOAP11_PROTOCOL)) {
+                soapFault.setFaultCode("soapenv:Server");
+            } else if (soapProtocolVersion.equals(APIConstants.SOAP12_PROTOCOL)) {
+                soapFault.setFaultCode("soapenv:Receiver");
+            }
+            soapFault.setFaultString(message);
+            soapFault.addDetail().addTextNode(code + ":" + description);
+            return envelope.toString();
+        } catch (Exception e) {
+            log.error("Error while creating the SOAP fault message.",
+                    ErrorDetails.errorLog(LoggingConstants.Severity.MINOR, 7101), e);
```

Comment:
```suggestion
            log.error("Error while creating the SOAP fault message. {}",
                    ErrorDetails.errorLog(LoggingConstants.Severity.MINOR, 7101), e.getMessage());
```

Reply:
fixed in https://github.com/wso2/product-microgateway/pull/2972/commits/5c91048f9fa451e4b2f2d91dac704d3f6b755cd3

Reply:
Since we are updating e to e.getMessage() it will be considered as a string. Therefore, another {} is needed.

Reply:
The `ErrorDetails.errorLog()` part is not considered as parameter here, its used by log4j2 to figure out the error code and details. So I think one {} is enough.

Reply:
The error log method signature is as follows.

```
public abstract void error(String message, Object p0 )
```
Every time a param gets added it gets updated to something like this.

```
public abstract void error(String message, Object p0, Object p1 )
```
You can check by hovering on to the method in the IDE. And yes, the `ErrorDetails.errorLog()` figures out the error code and details but it also is treated as another appended object. We need two `{} {}`. Can confirm this by triggering the log too.

Reply:
I tested the behavior by triggering this log statement. It's working as expected with single `{}`. 
Anyway, I think  `ErrorDetails.errorLog()` part is used to figure out the error code and severity level to be printed along with the log statement. It is not like embedding parameters to log message, error code and severity is printed with log statement separately.

Reply:
Thanks for testing out. +1